htmresearch/frameworks/clustering/sdr_clustering.py

```diff
@@ -0,0 +1,213 @@
+    return self._points
+
+
+  def size(self):
+    return len(self._points)
+
+
+  def merge(self, cluster):
+    for point in cluster.getPoints():
+      self.add(point)
+
+
+
+class Clustering(object):
+  def __init__(self,
+               mergeThreshold,
+               anomalousThreshold,
+               stableThreshold,
+               fistClusterId=0):
+
+    # Clusters
+    self.newCluster = Cluster(fistClusterId)
+    self.clusters = []
+
+    # Anomaly Score Thresholds
+    self.anomalousThreshold = anomalousThreshold  # to create new cluster
+    self.stableThreshold = stableThreshold  # to add point to cluster
+
+    # Cluster distance threshold
+    self.mergeThreshold = mergeThreshold
+
+
+  def findClosestClusters(self, cluster):
+    """
+    
+    :param cluster: 
+    :return: ordered list of clusters and their distances 
+    """
+    dists = []
+    for c in self.clusters:
+      d = clusterDist(c.getPoints(), cluster.getPoints())
+      dists.append((d, c))
+
+    return sorted(dists)
+
+
+  def addCluster(self, cluster):
+    self.clusters.append(cluster)
+    _LOGGER.debug("Cluster %s permanently added" % cluster.getId())
+
+
+  def addOrMergeCluster(self, cluster):
+
+    clusterDistPairs = self.findClosestClusters(cluster)
+    if len(clusterDistPairs) > 0:
+      notMerged = True
+      for clusterDistPair in clusterDistPairs: 
+        closestClusterToNewDist, closestClusterToNew = clusterDistPair
+  
+        if closestClusterToNewDist < self.mergeThreshold:
+          _LOGGER.debug("Cluster %s merged with cluster %s. Inter-cluster "
+                        "distance: %s" % (cluster.getId(),
+                                          closestClusterToNew.getId(),
+                                          closestClusterToNewDist))
+          closestClusterToNew.merge(cluster)
+          notMerged = False
+      if notMerged:
+        self.addCluster(cluster)
+    else:
+      self.addCluster(cluster)
+
+
+  def infer(self):
+    """
+    Inference: find the closest cluster to the new cluster.
+    """
+    clusterDistPairs = self.findClosestClusters(self.newCluster)
+    if len(clusterDistPairs) > 0:
+      distToCluster, predictedCluster = clusterDistPairs[0]
+      # Confidence of inference
+      meanClusterDist = np.mean([p[0] for p in clusterDistPairs])
+      if meanClusterDist > 0:
+        confidence = 1 - (distToCluster / meanClusterDist)
+      else:
+        if len(self.clusters) > 1:
+          raise ValueError("The mean distance can't be 0. Number of "
+                           "clusters: %s" % len(self.clusters))
+        else:
+          confidence = 1
+    else:
+      predictedCluster = None
+      confidence = -1
+
+    return predictedCluster, confidence
+
+
+  def cluster(self, sdrId, sdrValue, anomalyScore, trueLabel=None):
+
+    point = Point(sdrId, sdrValue, trueLabel)
+
+    # The data is anomalous
+    if self.anomalousThreshold <= anomalyScore:
+      if self.newCluster.size() > 0:
```

Comment:
Introduce a threshold here? To make sure that the new clusters are not too small